### ISSUE TITLE
Some dates are impossible to select and selection in some date ranges is very buggy in Safari

### ISSUE THREAD

**Issue by lolytxd** [vaadin/vaadin-date-picker-flow#126]
Vaadin version: 10.0.4

### Description
When I choose date 10/30/1995 everything is ok, but when I choose 10/29/1995 the date in the component is 10/27/1995. Every chosen date before 10/30/1995 have this bug.

### Actual outcome
Before clicking on 10/29/1995:
![image](https://user-images.githubusercontent.com/11396254/44725920-d484db00-aad6-11e8-87f2-7582cafa7ab3.png)

After clicking on 10/29/1995:
![image](https://user-images.githubusercontent.com/11396254/44725981-f8e0b780-aad6-11e8-9264-e2e216ff9f5f.png)

Date in field is 10/27/1995 but date selected in the date chooser is 10/28/1995. Date 10/29/1995 is not possible to be chosen.

### Steps to reproduce
Go to [https://vaadin.com/components/vaadin-date-picker/java-examples](https://vaadin.com/components/vaadin-date-picker/java-examples) in Firefox, and in first date picker choose date 10/29/1995, or write it, and then click enter.

### Browsers Affected
- [x] Firefox 61.0.1 x64

**Comment by Haprog** [comment]
@lolytxd Can you still reproduce this issue on some version of Firefox on https://vaadin.com/components/vaadin-date-picker/java-examples ?

I tried, but can't reproduce this at least on Firefox 67.0.4 (64-bit) or Firefox 60.7.2esr (64-bit) on macOS 10.14.5.

This seems to be the same issue as the recently reported #189 but it seems to be only happening on Safari anymore. At least I was only able to reproduce this on Safari.

**Comment by Haprog** [comment]
I closed #189 as a duplicate of this. Here is a copy of the most relevant info from #189 regarding this issue:

> @michel-mercera:
> Hi, I've made a screencast to show the behaviour. This is on Safari Version 12.1.1 (14607.2.6.1.1):
> [datepicker-vaadin-ex.mov.zip](https://github.com/vaadin/vaadin-date-picker/files/3338735/datepicker-vaadin-ex.mov.zip)

> @Haprog:
> I have the problem if I quickly scroll to some dates between 1960-1990 and then select a date by mouse, I can see that the date rendered in the text field is incorrect and also the "Day: " mentioned below is different from selected and from what is shown in input.
> 
> I also noticed if in this case right after selecting via mouse, you also click few times in the text input, the date changes to one date previous every time you click once more in the input.
>
> E.g. in Safari if I reload the page and then paste "6/1/1966" into the date picker and hit enter, then the date picker will show "5/27/1966" and below it will show "Day: 28".
> 
> ![image](https://user-images.githubusercontent.com/75155/60344278-db160480-99be-11e9-9e83-464790bd3368.png)
>
> Just verified that this issue still happens (in Safari) with latest version of `vaadin-date-picker-flow` (master, 2.0.1) when running the Java demos locally.
>
> The issue seems to be weirdly only in some seemingly random date ranges. For example for me selecting any date between "9/27/1993" – "10/31/1993" will cause the issue (and then clicking will go back one date until it hits "9/26/1993" which works) but selecting any of several dates immediately before or after that range works fine.

**Comment by ZheSun88** [comment]
This reminds me of an old strange ticket we had for date-picker. https://github.com/vaadin/vaadin-date-picker-flow/issues/128

The user has different locale settings and tried those on mobile devices. 
But we cannot reproduce it by any device we have, even on the same locale settings, so we decided to close it at that time. From the discussion here, it might be a valid faulty case. 

according to what I know, the connector doesn't do anything with the date, the regex will only filter the separators in the date format.

**Comment by Haprog** [comment]
@ZheSun88 That seems like the same issue too, but I can reproduce this at least now with desktop Safari 12 and also on iOS Safari in Simulator without changing any locale configuration (just by using the first demo "Simple date picker").

I just tested this affects also at least iOS Safari on iOS 10.3.1, 11.3 and 12.2.

**Comment by Haprog** [comment]
I updated the issue title to better describe the issue as we know it atm.

**Comment by Haprog** [comment]
The issue seems to be incorrect handling of timezone or DST for some dates or years in Safari at least on some timezones (depending on the user's location which is set in OS region settings).

`new Date()` constructor in Safari seems to detect incorrectly whether DST is in effect or not on some specific date/year combinations at least when user's OS region is set to e.g. Finland. And in these cases methods like `new Date().toLocaleDateString()` and `new Intl.DateTimeFormat().format()` return incorrect results.

Here are some examples:
```javascript
// With macOS "System Preferences" > "Language & Region" > "Region: Finland"

new Date(1966, 5, 1); // Safari: Wed Jun 01 1966 00:00:00 GMT+0300 (EEST), Chrome: Wed Jun 01 1966 00:00:00 GMT+0200 (Eastern European Summer Time), Firefox: 1966-05-31T22:00:00.000Z
new Date(2019, 5, 1); // Safari: Sat Jun 01 2019 00:00:00 GMT+0300 (EEST), Chrome: Sat Jun 01 2019 00:00:00 GMT+0300 (Eastern European Summer Time), Firefox: 2019-05-31T21:00:00.000Z

(new Date(1966, 5, 1)).toLocaleDateString('en-US'); // Safari: "5/31/1966", Chrome/Firefox: "6/1/1966"
(new Date(2019, 5, 1)).toLocaleDateString('en-US'); // Safari: "6/1/2019", Chrome/Firefox: "6/1/2019"

(new Date(1966, 5, 1)).toLocaleDateString('en-US', { hour: "numeric" }); // Safari: "5/31/1966, 11 PM", Chrome/Firefox: "6/1/1966, 12 AM"
(new Date(2019, 5, 1)).toLocaleDateString('en-US', { hour: "numeric" }); // Safari: "6/1/2019, 12 AM", Chrome/Firefox: "6/1/2019, 12 AM"

// Solution: use `Date.UTC` and `timeZone: 'UTC'` on `toLocaleDateString` to avoid timezone and DST conversions
(new Date(Date.UTC(1966, 5, 1))).toLocaleDateString('en-US', { hour: "numeric", timeZone: "UTC" }); // Safari: "6/1/1966, 12 AM", Chrome/Firefox: "6/1/1966, 12 AM"
(new Date(Date.UTC(2019, 5, 1))).toLocaleDateString('en-US', { hour: "numeric", timeZone: "UTC" }); // Safari: "6/1/2019, 12 AM", Chrome/Firefox: "6/1/2019, 12 AM"
```

`toLocaleDateString()` is used in two places in the `datepickerConnector.js`. The major issue being this part:
```javascript
datepicker.i18n.formatDate = function (date) {
  let rawDate = new Date(date.year, date.month, date.day);
  return cleanString(rawDate.toLocaleDateString(locale));
};
```

**Comment by Haprog** [comment]
This fix has been released in vaadin-date-picker-flow 1.0.7 (Vaadin 10), 1.3.1 (Vaadin 13) and 2.0.2 (Vaadin 14).
